### PR TITLE
fix(dcutr): exchange address candidates

### LINKED ISSUE #4496: dcutr: consider more than just external addresses
## Description

At the moment, the `dcutr::Behaviour` only sends _external addresses_ to the remote. See https://github.com/libp2p/rust-libp2p/blob/411a0495c5baa8dd714385a68871118ac98bbb61/protocols/dcutr/src/behaviour_impl.rs#L202 and https://github.com/libp2p/rust-libp2p/blob/411a0495c5baa8dd714385a68871118ac98bbb61/protocols/dcutr/src/behaviour_impl.rs#L103-L110

I'd argue this is kind of flawed. We don't need hole punching if we have external addresses. Instead, we should likely gather and use external address _candidates_ as well as allow the user to add their own candidates to the this list.

Likely, this is a left-over from before we landed https://github.com/libp2p/rust-libp2p/pull/3954.

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

In a serious of refactorings, we seem to have introduced a bug where we where exchanged the _external_ addresses of our node as part of `libp2p-dcutr`. This is ironically quite pointless. If we have external addresses, then there is no need for hole-punching (i.e. DCUtR).

Instead of gathering external addresses, we use an LRU cache to store our observed addresses. Repeatedly observed addresses will be tried first which should increase the success rate of a hole-punch.

Resolves: #4496.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/dcutr/src/behaviour.rs [hunk 1]
```diff
@@ -373,13 +368,50 @@ impl NetworkBehaviour for Behaviour {
...
+            me,
+        }
+    }
+
+    fn add(&mut self, address: Multiaddr) {
+        if is_relayed(&address) {
+            return;
+        }
+
+        let mut addr = address.clone();
```
Comment (mxinden) [inline protocols/dcutr/src/behaviour.rs]: Why is the `clone` necessary? `add` already takes an owned `Multiaddr`.

**Review by mxinden — APPROVED** [review]
Thanks for the fix and the test.

One comment, otherwise looks good to me.

File: protocols/dcutr/src/behaviour.rs (hunk 1 above)
Comment (thomaseizinger) [reply protocols/dcutr/src/behaviour.rs]: Let me investigate.

**Comment by thomaseizinger** [comment]
The dcutr example currently does not work. I attempted to fix this in https://github.com/libp2p/rust-libp2p/pull/4672.

File: examples/dcutr/src/main.rs [hunk 2]
```diff
@@ -157,7 +157,6 @@ async fn main() -> Result<(), Box<dyn Error>> {
                     ..
                 })) => {
                     info!("Relay told us our public address: {:?}", observed_addr);
-                    swarm.add_external_address(observed_addr);
```
Comment (mxinden) [inline examples/dcutr/src/main.rs:160]: Why remove this line? How else should the dcutr client discover its external address?

File: protocols/dcutr/src/behaviour.rs [hunk 3]
```diff
@@ -373,13 +368,48 @@ impl NetworkBehaviour for Behaviour {
...
+/// More common observations mean a more likely stable port-mapping and thus a higher chance of a successful hole-punch.
+struct Candidates {
+    inner: LruCache<Multiaddr, ()>,
+    me: PeerId,
+}
+
+impl Candidates {
+    fn new(me: PeerId) -> Self {
+        Self {
+            inner: LruCache::new(NonZeroUsize::new(100).expect("100 > 0")),
```
Comment (mxinden) [inline protocols/dcutr/src/behaviour.rs]:
```suggestion
            inner: LruCache::new(NonZeroUsize::new(20).expect("20 > 0")),
```

In which scenario would you expect a node to have more than say 10 external addresses? The DCUtR exchange is latency sensitive. Can 100 addresses be problematic, i.e. delay the address sending process?

File: examples/dcutr/src/main.rs (hunk 2 above)
Comment (thomaseizinger) [reply examples/dcutr/src/main.rs:160]: Why does it need its "external" address? We are sending over address candidates now which the dcutr behaviour learns automatically :)

Comment (thomaseizinger) [reply examples/dcutr/src/main.rs:160]: That is kind of the whole point of this PR, isn't it?

Comment (mxinden) [reply examples/dcutr/src/main.rs:160]: Ah, yes, yes, yes. I forgot that `libp2p-identify` emits the `ToSwarm::NewExternalAddrCandidate` :facepalm:

File: examples/dcutr/src/main.rs [hunk 4]
```diff
@@ -157,7 +157,6 @@ async fn main() -> Result<(), Box<dyn Error>> {
                     ..
                 })) => {
                     info!("Relay told us our public address: {:?}", observed_addr);
```
Comment (mxinden) [inline examples/dcutr/src/main.rs]:
```suggestion
                    info!("Relay told us our observed address: {:?}", observed_addr);
```

While we are at it. For future Max :P thinking like the past Max in the previous review.

**Review by mxinden — APPROVED** [review]

File: protocols/dcutr/Cargo.toml [hunk 5]
```diff
@@ -3,7 +3,7 @@ name = "libp2p-dcutr"
 edition = "2021"
 rust-version = { workspace = true }
 description = "Direct connection upgrade through relay"
-version = "0.11.0"
+version = "0.11.1"
```
Comment (thomaseizinger) [inline protocols/dcutr/Cargo.toml]:
```suggestion
version = "0.11.0"
```

File: Cargo.lock [hunk 6]
```diff
@@ -2482,7 +2482,7 @@ dependencies = [
 
 [[package]]
 name = "libp2p-dcutr"
-version = "0.11.0"
+version = "0.11.1"
```
Comment (thomaseizinger) [inline Cargo.lock]:
```suggestion
version = "0.11.0"
```

File: Cargo.toml [hunk 7]
```diff
@@ -76,7 +76,7 @@ libp2p-allow-block-list = { version = "0.3.0", path = "misc/allow-block-list" }
 libp2p-autonat = { version = "0.12.0", path = "protocols/autonat" }
 libp2p-connection-limits = { version = "0.3.0", path = "misc/connection-limits" }
 libp2p-core = { version = "0.41.0", path = "core" }
-libp2p-dcutr = { version = "0.11.0", path = "protocols/dcutr" }
+libp2p-dcutr = { version = "0.11.1", path = "protocols/dcutr" }
```
Comment (thomaseizinger) [inline Cargo.toml]:
```suggestion
libp2p-dcutr = { version = "0.11.0", path = "protocols/dcutr" }
```

**Comment by thomaseizinger** [comment]
Let's merge https://github.com/libp2p/rust-libp2p/pull/4549/ before this one.

**Pull request merged**